### ISSUE TITLE
[Question] is there a pattern for refreshing a page until something appears?

### ISSUE THREAD

**Issue by SteveALee** [microsoft/playwright#11703]
I can't figure out out the best way to handle items on a page that appear only when the page is refreshed.

I know the triggering action as it's part of the test. I could wait and then reload - but that will be flaky.

I could loop with a timeout I guess but is there a better way?

FYI the exact scenario is a form submission which creates a GitHub Pull Request via a webhook. The PR page on GitHub does not update so needs a reload so we can test for the new PR. System delays make the exact timing indeterminate.

Thanks

**Comment by aslushnikov** [comment]
@SteveALee this sounds like a good usecase to use [Github API to list pull requests](https://docs.github.com/en/rest/reference/pulls#list-pull-requests). Note that you can hit this API using Playwright's new [fetch api](https://playwright.dev/docs/api/class-apirequestcontext)

**Comment by SteveALee** [comment]
Interesting idea, but won't I still need to poll so the basic issue of how to reliably code in a playwright friendly way remains?

**Comment by aslushnikov** [comment]
> Interesting idea, but won't I still need to poll so the basic issue of how to reliably code in a playwright friendly way remains?

Yes, you would still need to poll. I don't think there's any way around this since we're trying to check the state of a remote system that doesn't provide notifications when its state changes.

**Comment by tymfear** [comment]
@aslushnikov I was recently wondering about kinda the same thing - sometimes we need to have some kind of retry logic, e.g. waiting for element to appear or retrying API call until it returns something. So how about some kind of waitFor... function that will be retrying until it returns true? Something like [waitForFunction](https://playwright.dev/docs/api/class-frame#frame-wait-for-function), but rather in Node context (or whatever language and runtime is used) than in browser context?
E.g.
```typescript
await page.waitForCondition(async () => {
  await page.reload();
  return page.locator('some locator').isVisible()
},  { timeout: 60 * 1000 })
```

Maybe it would be also useful to retry if some exception is throw in that function, kinda
```typescript
await page.waitForCondition(async () => {
  const value = Math.random();
  if (value > 0.7) {
   throw new Error('some error')
 } else {
   return value >= 0.3
}
},  { timeout: 60 * 1000 })
```
So that it will keep retrying either if error was thrown or if false was returned.

**Comment by SteveALee** [comment]
I think there's a class of wait tests needed for end to end testing which aren't covered by browsing to a url and then waiting for page updates (ie not SPA with realtime push updates). 

These are not obviously covered by playwright yet 

* resource created, eg http 404 to 200
* resource deleted, eg http 200 to 404
* page needs a refresh to update

**Comment by SteveALee** [comment]
@tymfear yes I scoured the docs for such a possibility  we have page.reload and page.waitForResponse and as its just javascript we can either use an imperative loop or better, take a more functional approach as you suggest. Best if provided for us though. :)

**Comment by SteveALee** [comment]
Finally a related issue is  I want to hide the slow warning as I've added a 20s delay to allow for full end to end delays. 

"Slow test file: [firefox] › form-test.spec.mjs (29s)
  Consider splitting slow test files to speed up parallel execution"

For now I appended "slow test" to the test description.

thanks

**Comment by SteveALee** [comment]
This seems related to #11736 - wait for events that are not triggered by playwright, plus ability to perform an action to trigger the even (eg reload)

**Comment by SteveALee** [comment]
yes, that will work. ;)

**Comment by soniaM7** [comment]
TypeError: page.waitForCondition is not a function

getting this error with --> await page.waitForCondition(async () => {    
this function

**Comment by rohitbonigala** [comment]
TypeError: page.waitForCondition is not a function

getting this error with --> await page.waitForCondition(async () => {
this function

**Comment by telboy007** [comment]
^^^ waitForCondition was a suggestion as how to handle this, not an actual function you can use (unfortunately).
